src/ape_http/providers.py

```diff
@@ -1,15 +1,111 @@
-from typing import Iterator
+import shutil
+from pathlib import Path
+from typing import Any, Dict, Iterator, List, Optional, Union
+from urllib.parse import urlparse
 
+from eth_typing import HexStr
+from geth import DevGethProcess, LoggingMixin  # type: ignore
+from requests.exceptions import ConnectionError
+from requests.exceptions import ConnectionError as RequestsConnectionError
 from web3 import HTTPProvider, Web3
+from web3.exceptions import ContractLogicError as Web3ContractLogicError
 from web3.gas_strategies.rpc import rpc_gas_price_strategy
 from web3.middleware import geth_poa_middleware
+from web3.types import RPCEndpoint
 
-from ape.api import ProviderAPI, ReceiptAPI, TransactionAPI
+from ape.api import ReceiptAPI, TransactionAPI, TransactionStatusEnum
 from ape.api.config import ConfigItem
-from ape.exceptions import ProviderError, TransactionError, VirtualMachineError
-from ape.utils import get_tx_error_from_web3_value_error
+from ape.api.providers import TestProviderAPI
+from ape.exceptions import (
+    ContractLogicError,
+    OutOfGasError,
+    ProviderError,
+    RPCError,
+    TransactionError,
+    VirtualMachineError,
+)
+from ape.logging import logger
+from ape.utils import generate_dev_accounts, get_gas_estimation_revert_error_message
 
-DEFAULT_SETTINGS = {"uri": "http://localhost:8545"}
+DEFAULT_SETTINGS = {"uri": "http://localhost:8545", "chain_name": "devchain"}
+
+
+class EphemeralGeth(LoggingMixin, DevGethProcess):
+    """
+    A developer-configured geth that only exists until disconnected.
+    """
+
+    def __init__(
+        self,
+        chain_name,
+        base_directory: Path,
+        hostname: str,
+        port: int,
+        chain_id: int = 1337,
+        initial_balance: Union[str, int] = "10000000000000000000000",
+        accounts: List = None,
+    ):
+        self.data_dir = base_directory / chain_name
+        self._hostname = hostname
+        self._port = port
+
+        accounts = accounts or generate_dev_accounts()
+        genesis_data: Dict = {
+            "coinbase": "0x0000000000000000000000000000000000000000",
+            "config": {
+                "chainId": chain_id,
+                "gasLimit": 0,
+                "homesteadBlock": 0,
+                "difficulty": "0x0",
+                "eip150Block": 0,
+                "eip155Block": 0,
+                "eip158Block": 0,
+                "byzantiumBlock": 0,
+                "constantinopleBlock": 0,
+                "petersburgBlock": 0,
+                "istanbulBlock": 0,
+                "berlinBlock": 0,
+                "londonBlock": 0,
+                "ethash": {},
+            },
+            "alloc": {a.address: {"balance": str(initial_balance)} for a in accounts},
+        }
+        geth_cmd_args = {
+            "rpc_addr": hostname,
+            "rpc_port": str(port),
```

Comment:
As soon as the next py-geth is released, we won't have to call `str()` on the ints, we won't need to specify and `allow_insecure_unlock`